onnxruntime/core/mlas/lib/q4_dq.cpp

```diff
@@ -294,3 +294,584 @@
+
+                const int meta_row = r / QuantBlk::kRow;
+                const int meta_col = c / QuantBlk::kColumn;
+
+                if constexpr (Columnwise) {
+                    static_assert(ThreadBlk::kColumn == 1, "Internal computation error!");
+                    int row_idx = r;
+                    for (int kpack = 0; kpack < BitsTraits<qbits>::kPackSize && row_idx < r_end;
+                         kpack++) {
+                        float min = std::numeric_limits<float>::max();
+                        float max = -min;
+                        for (int idx = 0; idx < QuantBlk::kRow && row_idx < r_end;
+                             idx++, row_idx++) {
+                            const float v = static_cast<float>(src[row_idx * leadingDimension + c]);
+                            if (v < min) min = v;
+                            if (v > max) max = v;
+                        }
+
+                        // store scale and zero point at matrix position
+                        const int32_t meta_idx = meta_col * row_blks + meta_row + kpack;
+                        if (zero_points == nullptr) {
+                            range2scale<ElementT, qbits>(min, max, scales[meta_idx]);
+                        } else {
+                            range2scalezp<ElementT, qbits>(min, max, scales[meta_idx],
+                                                           zp_bytes[kpack]);
+                        }
+                    }
+
+                } else {  // Row-wise
+                    static_assert(ThreadBlk::kRow == BitsTraits<qbits>::kPackSize,
+                                  "Internal computation error!");
+                    for (int32_t i = r; i < r_end; ++i) {
+                        float min = std::numeric_limits<float>::max();
+                        float max = -min;
+                        for (int32_t j = c; j < c_end; ++j) {
+                            const float v = static_cast<float>(src[i * leadingDimension + j]);
+                            if (v < min) min = v;
+                            if (v > max) max = v;
+                        }
+                        // store scale and zero point at matrix position (i/QuantBlk::kRow,
+                        // j/QuantBlk::kColumn) it's packed as column major, so the index is (column
+                        // * row_blks + row) over all it's (j/QuantBlk::kColumn) * row_blks +
+                        // (i/QuantBlk::kRow)
+                        const int32_t meta_idx = meta_col * row_blks + (i / QuantBlk::kRow);
+                        if (zero_points == nullptr) {
+                            range2scale<ElementT, qbits>(min, max, scales[meta_idx]);
+                        } else {
+                            range2scalezp<ElementT, qbits>(min, max, scales[meta_idx],
+                                                           zp_bytes[i-r]);
+                        }
+                    }
+                }
```

Comment:
I'm thinking it is good to extend the blockwise to general. We quantize with a 2D size [block_size_row, block_size_col]. Then columnwise and rowwise will be special case: columnwise is [block_size_row, 1] and rowwise is [1, block_size_col]. And then those code can be combined

Reply:
For a matrix [row, col] with block size [block_row, block_col], the quantized matrix will be stored as [ceil(col / block_col)][ceil(row/ block_row)][block_row * block_col /2] if using transpose

Reply:
That might be promising too much. It's quite a bit of more corner cases and harder to tests. Unless there is a clear indication of real world needs, let's delay this feature.

Reply:
I considered your suggestion again, and after relaxing the quantized packed shape to multiple of whole blocks, now I can remove those special case branches as you suggested.

To offer this as an interface to the user, however, we need to add more tests. Tests fail mostly happens when a certain dimension is not whole blocks and/or is an odd number.  It's very cumbersome to support these odd cases. Since LLM matrix shapes are often multiple of 64, can we restrict this feature to shapes only multiple of 32?